[BUG] KoP commit offset failed when doing unload

**Describe the bug**

Currently, when the bundle is to unload, the GroupMetadataManager's producer will be closed, when the KoP receives a commit offset request, then this request will fail with an unknown error. 

Client-side log:
```
7:29:51.440 [pool-3-thread-1] ERROR io.streamnative.tests.pulsar.chaos.test.DurationTestBase - Failed to acknowledge message. offset: 42, receiveIndex: 339, lastReceiveIndex: -1, errMsg: Unexpected error in commit: The server experienced an unexpected error when processing the request.
org.apache.kafka.common.KafkaException: Unexpected error in commit: The server experienced an unexpected error when processing the request.
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator$OffsetCommitResponseHandler.handle(ConsumerCoordinator.java:1260) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator$OffsetCommitResponseHandler.handle(ConsumerCoordinator.java:1163) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator$CoordinatorResponseHandler.onSuccess(AbstractCoordinator.java:1177) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator$CoordinatorResponseHandler.onSuccess(AbstractCoordinator.java:1152) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.RequestFuture$1.onSuccess(RequestFuture.java:206) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.RequestFuture.fireSuccess(RequestFuture.java:169) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.RequestFuture.complete(RequestFuture.java:129) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient$RequestFutureCompletionHandler.fireCompletion(ConsumerNetworkClient.java:602) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.firePendingCompletedRequests(ConsumerNetworkClient.java:412) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:297) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:236) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:215) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.commitOffsetsSync(ConsumerCoordinator.java:1005) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.KafkaConsumer.commitSync(KafkaConsumer.java:1490) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.KafkaConsumer.commitSync(KafkaConsumer.java:1390) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.KafkaConsumer.commitSync(KafkaConsumer.java:1347) ~[kafka-clients-2.8.0.jar:?]
	at io.streamnative.tests.pulsar.chaos.test.KoPMessagingAtLeastOnceTest.lambda$consumeMsg$0(KoPMessagingAtLeastOnceTest.java:251) ~[test-classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
```

```
07:29:51.341 [Thread-8850] ERROR io.streamnative.pulsar.handlers.kop.coordinator.group.GroupMetadataManager - Offset commit {persistent://kop/semantic/at-least-once-akhwz-0=[OffsetMetadata[40,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-1=[OffsetMetadata[29,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-2=[OffsetMetadata[41,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-3=[OffsetMetadata[40,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-4=[OffsetMetadata[31,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-5=[OffsetMetadata[38,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-6=[OffsetMetadata[43,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-7=[OffsetMetadata[40,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-8=[OffsetMetadata[42,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252]} from group group1, consumer DemoKafkaOnPulsarConsumerSSL-2b55510f-d7e3-463d-ad32-cdc6a52bab3c with generation 6 failed when appending to log due to 
java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException$AlreadyClosedException: The producer chaos-pulsar-28-7-34 of the topic public/__kafka/__consumer_offsets-partition-10 was already closed when closing the producers
	at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:367) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:376) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1019) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088) ~[?:?]
	at org.apache.pulsar.client.impl.ProducerImpl$1.sendComplete(ProducerImpl.java:347) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl$OpSendMsg.sendComplete(ProducerImpl.java:1291) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl.lambda$failPendingMessages$18(ProducerImpl.java:1827) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at java.util.ArrayDeque.forEach(ArrayDeque.java:889) ~[?:?]
	at org.apache.pulsar.client.impl.ProducerImpl$OpSendMsgQueue.forEach(ProducerImpl.java:1370) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl.failPendingMessages(ProducerImpl.java:1817) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl.closeAndClearPendingMessages(ProducerImpl.java:964) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl.closeAsync(ProducerImpl.java:931) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at io.streamnative.pulsar.handlers.kop.coordinator.group.GroupMetadataManager.lambda$null$74(GroupMetadataManager.java:1169) ~[C44vRLkVGY-jWrCmMdzE9A/:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642) [?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: org.apache.pulsar.client.api.PulsarClientException$AlreadyClosedException: The producer chaos-pulsar-28-7-34 of the topic public/__kafka/__consumer_offsets-partition-10 was already closed when closing the producers
	at org.apache.pulsar.client.impl.ProducerImpl.closeAndClearPendingMessages(ProducerImpl.java:961) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	... 5 more
```

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupMetadataManager.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupMetadataManager.java
@@ -584,6 +584,7 @@ public class GroupMetadataManager {
                         filteredOffsetMetadata, group.groupId(), consumerId, group.generationId(), cause);
 
                 if (cause.getCause() instanceof PulsarClientException.AlreadyClosedException) {
+                    this.offsetsProducers.remove(partitionFor(group.groupId()));
                     return Errors.NOT_COORDINATOR;
                 }
                 return Errors.UNKNOWN_SERVER_ERROR;

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupMetadataManager.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupMetadataManager.java
@@ -82,6 +82,7 @@ import org.apache.pulsar.client.api.Message;
 import org.apache.pulsar.client.api.MessageId;
 import org.apache.pulsar.client.api.Producer;
 import org.apache.pulsar.client.api.ProducerBuilder;
+import org.apache.pulsar.client.api.PulsarClientException;
 import org.apache.pulsar.client.api.Reader;
 import org.apache.pulsar.client.api.ReaderBuilder;
 import org.apache.pulsar.client.impl.MessageIdImpl;
@@ -582,6 +583,9 @@ public class GroupMetadataManager {
                                 + " when appending to log due to ",
                         filteredOffsetMetadata, group.groupId(), consumerId, group.generationId(), cause);
 
+                if (cause.getCause() instanceof PulsarClientException.AlreadyClosedException) {
+                    return Errors.NOT_COORDINATOR;
+                }
                 return Errors.UNKNOWN_SERVER_ERROR;
             })
             .thenApplyAsync(errors -> offsetMetadata.entrySet()

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java
@@ -54,7 +54,10 @@ import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.stream.Collectors;
 import java.util.stream.LongStream;
@@ -74,6 +77,7 @@ import org.apache.kafka.clients.producer.KafkaProducer;
 import org.apache.kafka.clients.producer.ProducerConfig;
 import org.apache.kafka.clients.producer.ProducerRecord;
 import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.KafkaException;
 import org.apache.kafka.common.Node;
 import org.apache.kafka.common.TopicPartition;
 import org.apache.kafka.common.config.ConfigResource;
@@ -106,6 +110,7 @@ import org.apache.kafka.common.serialization.IntegerSerializer;
 import org.apache.kafka.common.serialization.StringSerializer;
 import org.apache.kafka.common.utils.Time;
 import org.apache.pulsar.client.admin.PulsarAdminException;
+import org.apache.pulsar.client.api.PulsarClientException;
 import org.apache.pulsar.common.allocator.PulsarByteBufAllocator;
 import org.apache.pulsar.common.naming.TopicName;
 import org.apache.pulsar.common.policies.data.AutoTopicCreationOverride;
@@ -1173,6 +1178,65 @@ public class KafkaRequestHandlerTest extends KopProtocolHandlerTestBase {
             conf.setAllowAutoTopicCreation(original);
             admin.namespaces().removeAutoTopicCreation("public/default");
         }
+    }
+
+    @Test(timeOut = 30000)
+    public void testCommitOffsetRetryWhenProducerClosed()
+            throws ExecutionException, InterruptedException, PulsarAdminException {
+        String topic = "testCommitOffsetRetryWhenProducerClosed";
+        String groupId = "test-group";
+        admin.topics().createPartitionedTopic(topic, 1);
+        int numMessages = 10;
+        @Cleanup
+        final KafkaProducer<String, String> producer = new KafkaProducer<>(newKafkaProducerProperties());
+        for (int i = 0; i < numMessages; i++) {
+            producer.send(new ProducerRecord<>(topic, "value")).get();
+        }
+
+        @Cleanup
+        final KafkaConsumer<String, String> consumer = new KafkaConsumer<>(newKafkaConsumerProperties(groupId));
+        consumer.subscribe(Collections.singleton(topic));
+
+        int fetchMessages = 0;
+
+        // Need open another thread to clear the offset producers
+        Executors.newSingleThreadExecutor().submit(() -> {
+            try {
+                TimeUnit.SECONDS.sleep(5);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+            handler.getGroupCoordinator().getOffsetsProducers().clear();
+        });
+
+        // Make sure only close once.
+        final AtomicBoolean flag = new AtomicBoolean(true);
+
+        while (fetchMessages < numMessages) {
+            try {
+                ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(1000));
+                records.forEach(record -> {
+                    record.offset();
+                    record.partition();
+                    consumer.commitSync();
+                    if (flag.get()) {
+                        handler.getGroupCoordinator().getOffsetsProducers().values()
+                                .forEach(producerCompletableFuture -> {
+                            try {
+                                producerCompletableFuture.get().close();
+                            } catch (PulsarClientException | InterruptedException | ExecutionException e) {
+                                log.error("Close offset producer failed.");
+                            }
+                        });
+                        flag.set(false);
+                    }
+                });
+                fetchMessages += records.count();
+            } catch (KafkaException ex) {
+                fail("Should not have kafka exception.");
+            }
+        }
+        assertEquals(fetchMessages, numMessages);
     }
 
     private KafkaHeaderAndRequest createTopicMetadataRequest(List<String> topics, boolean allowAutoTopicCreation) {

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java
@@ -54,9 +54,7 @@ import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
-import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.stream.Collectors;
@@ -1199,16 +1197,6 @@ public class KafkaRequestHandlerTest extends KopProtocolHandlerTestBase {
 
         int fetchMessages = 0;
 
-        // Need open another thread to clear the offset producers
-        Executors.newSingleThreadExecutor().submit(() -> {
-            try {
-                TimeUnit.SECONDS.sleep(5);
-            } catch (InterruptedException e) {
-                e.printStackTrace();
-            }
-            handler.getGroupCoordinator().getOffsetsProducers().clear();
-        });
-
         // Make sure only close once.
         final AtomicBoolean flag = new AtomicBoolean(true);
 
@@ -1216,8 +1204,6 @@ public class KafkaRequestHandlerTest extends KopProtocolHandlerTestBase {
             try {
                 ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(1000));
                 records.forEach(record -> {
-                    record.offset();
-                    record.partition();
                     consumer.commitSync();
                     if (flag.get()) {
                         handler.getGroupCoordinator().getOffsetsProducers().values()
@@ -1233,7 +1219,8 @@ public class KafkaRequestHandlerTest extends KopProtocolHandlerTestBase {
                 });
                 fetchMessages += records.count();
             } catch (KafkaException ex) {
-                fail("Should not have kafka exception.");
+                log.error("Have kafka exception: ", ex);
+                throw ex;
             }
         }
         assertEquals(fetchMessages, numMessages);

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java
@@ -1184,7 +1184,7 @@ public class KafkaRequestHandlerTest extends KopProtocolHandlerTestBase {
     public void testCommitOffsetRetryWhenProducerClosed()
             throws ExecutionException, InterruptedException, PulsarAdminException {
         String topic = "testCommitOffsetRetryWhenProducerClosed";
-        String groupId = "test-group";
+        String groupId = "test-commit-offset-group";
         admin.topics().createPartitionedTopic(topic, 1);
         int numMessages = 10;
         @Cleanup

